### ISSUE TITLE
Some questions about the code

### ISSUE THREAD

**Issue by Baijiong-Lin** [ATP-1010/FederatedLLM#3]
Thanks for your excellent work and code. I have some questions about the code.

1. when ``stacking == True`` and ``zero_padding == True``?
https://github.com/ATP-1010/FederatedLLM/blob/8efe72d1a7c77456f5ab342efd6394732bfabe05/main.py#L201-L214
https://github.com/ATP-1010/FederatedLLM/blob/8efe72d1a7c77456f5ab342efd6394732bfabe05/fed_utils/model_aggregation.py#L33-L34

2. If my understanding is correct, after training, we need to save the merged model ckpt for FLoRA (since the local model (it is the pretrained model at the first round) is modified at the end of each round) while we only need to save the lora model for FedIT. But, it seems in this code, both methods save the final lora model only.

I hope you can help me to address these problems. Thank you.

**Comment by ATP-1010** [comment]
For the second question, I have update the code several days ago to make it only save the LoRA modules. You may try to update the code and find if the problem is solved. For the first question, let me double check my local code and give you feedback tomorrow.

**Comment by Baijiong-Lin** [comment]
Thank you for getting back to me so quickly. I have used the latest code, which is updated on Oct 12.

**Comment by ATP-1010** [comment]
I have made several updates today, please check the code again. I will keep trying to make the code clearer. There are still some difference between this code and my local code.

**Comment by ATP-1010** [comment]
BTW, thank you for the valuable issues.

**Comment by Baijiong-Lin** [comment]
Thanks for your timely update. But it still has some problems, like

https://github.com/ATP-1010/FederatedLLM/blob/f1c6519769d18a87ca03e995934fe1166b8f98ab/fed_utils/model_aggregation.py#L33-L34

Can the current code successfully reproduce the results in the paper?

**Comment by Baijiong-Lin** [comment]
Did you use this tinyllama in the experiment?

https://github.com/ATP-1010/FederatedLLM/blob/f1c6519769d18a87ca03e995934fe1166b8f98ab/download.py#L3

**Comment by ATP-1010** [comment]
You dont need to use this code, this is for some server that cannot download models automatically. Just list TinyLlama as your global model.

**Comment by Baijiong-Lin** [comment]
I need to download the model manually on my server. So could you provide the hf link for the tinyllama you used?

**Comment by ATP-1010** [comment]
TinyLlama/TinyLlama-1.1B-Chat-v1.0

**Comment by Baijiong-Lin** [comment]
I cannot reproduce the paper results by using the latest code. 

1. I ran the command ``python main.py --global_model ~/hf_ckpts/ckpts/TinyLlama--TinyLlama-1.1B-Chat-v1.0/ --data_path ./data_wiz --output_dir ./exp/FloRA-tinyllama-wiz-homo/ --num_communication_rounds 3 --local_num_epochs 1 --stacking True``.
After training, only the stacked lora model is saved on my machine (``adapter_model.bin`` in ``./exp/FloRA-tinyllama-wiz-homo/10/2``). Hence, it seems the second problem I raised initially still exists. 

3. I also try to reproduce the results of FedIT by running the command ``python main.py --global_model ~/hf_ckpts/ckpts/TinyLlama--TinyLlama-1.1B-Chat-v1.0/ --data_path ./data_wiz --output_dir ./exp/FedIT-tinyllama-wiz-homo/ --num_communication_rounds 3 --local_num_epochs 1 --stacking False``.
After training, I ran the commend ``lm_eval --model_args pretrained=~/hf_ckpts/ckpts/TinyLlama--TinyLlama-1.1B-Chat-v1.0/,peft=./exp/FedIT-tinyllama-wiz-homo/10/2/ --tasks mmlu --num_fewshot 5 --device cuda`` to evaluate the saved lora model. However, the accuracy of MMLU is only 25.25%.

**Comment by ATP-1010** [comment]
I uploaded an efficient MMLU test after each round. Please see if this and MT-bench can help you reproduce the results (Sometimes the harness gives unchanged numbers after fine-tuning). I am trying to save the full parameter separately in a fold final now. Please try to see if it solves your second issue.

**Comment by Baijiong-Lin** [comment]
Could you run the code and ensure it can reproduce the results? Actually, I have tried to save the whole model but it still has a low evaluation accuracy on MMLU. Or could you send me your local code by email? It's okay if the code is not clean, as long as the correct results can be obtained.

**Comment by ATP-1010** [comment]
Please use the new code with evaluation inside main.py:
        acc = global_evaluation(model, tokenizer, prompter, dev_data_path)
        print('Acc of Epoch', str(epoch), 'is:', acc)
        acc_list.append(acc)
Let me know if you can reproduce the Acc. I'm sorry that the local code is now modified for another project (this is why this code has some discrepancies with the FLoRA paper) that cannot be shared now. The Acc listed in the paper is not done by harness. Please also use the MT-bench to check if you can reproduce the performance in the paper, both baselines and ours.

**Comment by Baijiong-Lin** [comment]
The reproduced results (TinyLlama, Homo) are as follows (on ``mmlu_test_1444.jsonl`` / using ``lm_eval``). 

FLoRA: 36.31/25.40
FedIT: 33.06/24.97

There is still a big gap between my reproduced results and the paper's results. 

BTW, where is the source of the file ``mmlu_test_1444.jsonl``.

**Comment by ATP-1010** [comment]
I have changed the hyperparameters in main.py to fit the settings in the paper. Please try the new settings (lr, batch size, lora ranks are changed).
mmlu_test_1444.jsonl is selected from the full mmlu test set (10 times more samples) for efficient evaluation. I think here the main difference from lm harness is the evaluation codes. Harness uses probabilities, but our code detects the answers in the output. When using lm harness, I often find that the results before and after fine-tuning are almost unchanged, making it difficult to compare the effectiveness of our method against the baseline. If you are happy to try the full mmlu test dataset, I have uploaded it for your reference (mmlu_test_14042.jsonl).

**Comment by huskydj1** [comment]
I am encountering a similar but different problem with reproducing the results in the paper. Using TinyLlama/homogenous/wizard/MMLU, I have reproduced the following results (on mmlu_test_14042.jsonl / using lm_eval):

0.4133/0.2495

Strangely enough, the performance on all sets gets worse after every epoch. For instance, on mmlu_test_14042:

Epoch 0: 0.46782455591149263
Epoch 1: 0.4222263906230094
Epoch 2: 0.41330578892413367

Do you know what the reason might be?

**Comment by ATP-1010** [comment]
Looks like it is overfitting. It's very common in fine-tuning LLMs on some common datasets and benchmarks.
Could you please try to slightly adjust some hyperparameters? We also don't have optimal solutions if overfitting happens.

**Comment by hwan-sig** [comment]
Hi, may I ask you a question about the code?
When you trained with data_wiz, did you only change the `lora_dropout` parameter?
